src/drivers/mediatek/mt8195/afe-drv.c

```diff
@@ -0,0 +1,421 @@
+		return -EINVAL;
+	}
+	dai = &afe->dais[id];
+
+	*channel = dai->channel;
+	*rate = dai->rate;
+	*format = dai->format;
+
+	tr_info(&afedrv_tr, "dai:%d get: format:%d, rate:%d, channel:%d\n", id, *format, *rate,
+		*channel);
+
+	return 0;
+}
+
+/* TODO, IRQ common register name need use config? */
+int afe_irq_get_status(struct mtk_base_afe *afe, int id)
+{
+	return 0;
+}
+
+int afe_irq_clear(struct mtk_base_afe *afe, int id)
+{
+	return 0;
+}
+
+int afe_irq_config(struct mtk_base_afe *afe, int id, unsigned int rate, unsigned int period)
+{
+	struct mtk_base_afe_irq *irq = &afe->irqs[id];
+	unsigned int fs;
+
+	afe_reg_update_bits(afe, irq->irq_data->irq_cnt_reg,
+			    irq->irq_data->irq_cnt_maskbit << irq->irq_data->irq_cnt_shift,
+			    period << irq->irq_data->irq_cnt_shift);
+
+	/* set irq fs */
+	fs = afe->irq_fs(rate);
+
+	if (fs < 0)
+		return -EINVAL;
+
+	afe_reg_update_bits(afe, irq->irq_data->irq_fs_reg,
+			    irq->irq_data->irq_fs_maskbit << irq->irq_data->irq_fs_shift,
+			    fs << irq->irq_data->irq_fs_shift);
+	return 0;
+}
+
+/* TODO, for dma based scheduler*/
+int afe_irq_enable(struct mtk_base_afe *afe, int id)
+{
+	return 0;
+}
+
+int afe_irq_disable(struct mtk_base_afe *afe, int id)
+{
+	return 0;
+}
+
+struct mtk_base_afe *afe_get(void)
+{
+	return &mtk_afe;
+}
+
+int afe_probe(struct mtk_base_afe *afe)
+{
+	int ret = 0;
+	int i;
+	struct mtk_base_afe_platform *platform = &mtk_afe_platform;
+
+	/* mtk afe already init done */
+	if (afe->ref_count > 0) {
+		afe->ref_count++;
+		return ret;
+	}
+
+	afe->platform_priv = platform;
+	afe->base = platform->base_addr;
+	afe->memif_32bit_supported = platform->memif_32bit_supported;
+	afe->memif_dl_num = platform->memif_dl_num;
+
+	afe->base_end_offset = platform->base_end_offset;
+	afe->adsp2afe_addr = platform->adsp2afe_addr;
+	afe->afe2adsp_addr = platform->afe2adsp_addr;
+	afe->afe_fs = platform->afe_fs; /* must be */
+	afe->irq_fs = platform->irq_fs;
+	if (!afe->afe_fs)
+		return -EINVAL;
+	tr_dbg(&afedrv_tr, "afe_base:0x%x\n", afe->base);
+	/* TODO how to get the memif number, how to sync with dmac lib */
+	afe->memifs_size = platform->memif_size;
+	afe->memif = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+			     sizeof(struct mtk_base_afe_memif) * afe->memifs_size);
+	if (!afe->memif)
+		return -ENOMEM;
+
+	for (i = 0; i < afe->memifs_size; i++)
+		afe->memif[i].data = &platform->memif_datas[i];
+
+	/* TODO how to get the dai number, how to sync with dai lib*/
+	afe->dais_size = platform->dais_size;
+	afe->dais = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+			    sizeof(struct mtk_base_afe_dai) * afe->dais_size);
+	if (!afe->dais) {
+		ret = -ENOMEM;
+		goto err_alloc_memif;
+	}
+	/* TODO how to get the irq number */
+	afe->irqs_size = platform->irqs_size;
+	afe->irqs = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+			    sizeof(struct mtk_base_afe_irq) * afe->irqs_size);
```

Comment:
`SHARED` is ignored on single-core configurations. It's ok to use it in case you might in the future get multi-core configurations. Otherwise you could just use `SOF_MEM_ZONE_RUNTIME`